### PR TITLE
Add on the fly demux and speed up barcode matching.

### PR DESCRIPTION
**Pull request by jacarey**
### Description

The purpose of this PR is twofold:

1) Speed up the barcode matching code.
2) Add inline demux to `IlluminaBasecallsToSam` and `IlluminaBasecallsToFastq`

Sample input parameters are parsed in a central place now which can parse any one of the following types of files:
 

-       ExtractIlluminaBarcodes      BARCODE_FILE
-       IlluminaBasecallsToFastq     MULTIPLEX_PARAMS
-       IlluminaBasecallsToSam       LIBRARY_PARAMS

A single barcode extractor is used for all 3 programs to do demultiplexing and metrics collection/output.

Test data was changed for the following reason:

The old test data was using static barcode files that were created before `HAMMING` was introduced. In order to minimize the number of test data files needed the barcode files were regenerated using `HAMMING` and the test data was updated to reflect this change.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by jacarey** [comment]
Hmm I can't seem to get these tests to fail locally again. @gbggrant would you mind testing as well?

**Comment by jacarey** [comment]
@gbggrant I think it makes the most sense to concentrate on the other PRs right now and then once they are merged come back to this one after a rebase since it will be changed significantly by the others.

**Comment by jacarey** [comment]
@gbggrant This one is a little bit more substantive but it is ready for review.

**Review by jessicaway — APPROVED** [review]
Looks good overall, a few minor comments

**Review by gbggrant — APPROVED** [review]
Looks good to me.  A couple of stylistic notes and a question about a comment.

**Comment by gbggrant** [comment]
Hi @jacarey we've reviewed this one.  Just checking in to see if this is the highest priority PR of yours (I see you have another (https://github.com/broadinstitute/picard/pull/1646) but it's still in draft.

**Comment by jacarey** [comment]
> Hi @jacarey we've reviewed this one. Just checking in to see if this is the highest priority PR of yours (I see you have another (#1646) but it's still in draft.

@gbggrant I'm hoping to get to this early next week. I've moved 1646 from draft to PR. Thanks!

File: src/main/java/picard/illumina/ExtractIlluminaBarcodes.java [hunk 1]
```diff
@@ -255,27 +164,28 @@ protected int doWork() {
...
-        }
+        final BarcodeExtractor barcodeExtractor = createBarcodeExtractor();
+        final Set<IlluminaDataType> datatypes = (MINIMUM_BASE_QUALITY > 0) ?
+                new HashSet<>(Arrays.asList(IlluminaDataType.BaseCalls, IlluminaDataType.PF, IlluminaDataType.QualityScores)) :
+                new HashSet<>(Arrays.asList(IlluminaDataType.BaseCalls, IlluminaDataType.PF));
+        final List<PerTileBarcodeExtractor> extractors = new ArrayList<>();
+        for(Integer lane: LANE) {
+            IlluminaDataProviderFactory factory = new IlluminaDataProviderFactory(BASECALLS_DIR, lane, inputReadStructure, bclQualityEvaluationStrategy, datatypes);
+
+            // TODO: This is terribly inefficient; we're opening a huge number of files via the extractor constructor and we never close them.
```
[the remark this answers is not in the document]
Comment (jacarey) [reply src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:175]: Hmm this is an old comment. We are using a data provider so I don't think we can close them until we are done extraction.

[the remark this answers is not in the document]
Comment (jacarey) [reply src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:175]: Ah I think we can delay provider creation to just prior to use.

File: src/test/java/picard/illumina/parser/BinTdUtil.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ public static final String ltStr(final int lane, final int tile) {
                 makeCd(lane, tile, 1159, 2179, false,
                         new byte[]{G,T,T,A,G,C,A,C,A,G,A,T,A,T,T,G,G,A,T,G,A,G,T,G,A,A,A,A,A,A,A,A,A,T,T,T,T,T,T,T,T,T,A,T,T,T,T,T,C,T,A,A,A,T,A,C,T,T},
                         new byte[]{2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,28,28,28,32,32,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2},
-                        null),
+                        "AAAAAAAA"),
```
[the remark this answers is not in the document]
Comment (jacarey) [reply src/test/java/picard/illumina/parser/BinTdUtil.java:69]: If you look a the actual bcl data it doesn't match what is in the barcodes file. I'm guessing that data had been updated at some point after the barcodes files were generated. So we regenerated those files using ExtractIllumiunaBarcodes and now they are pulling out AAAAAAAA reads that didn't exist before.

**Comment by mjhipp** [comment]
Even without on the fly demux, this saves me a full day when demultiplexing a single NovaSeq lane. Would love to be able to use an official release with this feature!

File: src/main/java/picard/illumina/BasecallsConverter.java [hunk 3]
```diff
@@ -264,4 +318,13 @@ protected void interruptAndShutdownExecutors(ThreadPoolExecutorWithExceptions...
         throw new PicardException("Exceptions in tile processing. There were " + tasksRunning
                 + " tasks still running or queued and they have been cancelled. Errors: " + errorMessages);
     }
+
+    protected synchronized void updateMetrics(Map<String, BarcodeMetric> metrics, BarcodeMetric noMatch) {
+        if(barcodeExtractor != null) {
```
Comment (gbggrant) [inline src/main/java/picard/illumina/BasecallsConverter.java:323]:
```suggestion
        if (barcodeExtractor != null) {
```

File: src/main/java/picard/illumina/ExtractBarcodesProgram.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+                inputReadStructure,
+                MAX_NO_CALLS,
+                MAX_MISMATCHES,
+                MIN_MISMATCH_DELTA,
+                MINIMUM_BASE_QUALITY,
+                DISTANCE_MODE);
+    }
+
+    /**
+     * Validate that POSITION >= 1, and that all BARCODEs are the same length and unique
```
Comment (gbggrant) [inline src/main/java/picard/illumina/ExtractBarcodesProgram.java:121]: Is this comment right?

File: src/main/java/picard/illumina/ExtractBarcodesProgram.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+                for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+                    barcodeMetric.PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT =
+                            barcodeMetric.PF_PCT_MATCHES / bestPctOfAllBarcodeMatches;
+                }
+            }
+        }
+
+        // Calculate the normalized matches
+        if (totalPfReadsAssigned > 0) {
+            final double mean = (double) totalPfReadsAssigned / (double) barcodeToMetrics.values().size();
```
Comment (gbggrant) [inline src/main/java/picard/illumina/ExtractBarcodesProgram.java:218]: Can barcodeToMetrics ever be empty?

File: src/main/java/picard/illumina/ExtractBarcodesProgram.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            }).collect(Collectors.toList());
+
+            if (readStructure.sampleBarcodes.length() != validBarcodeColumns.size()) {
+                messages.add("Expected " + readStructure.sampleBarcodes.length() + " valid barcode columns, but found " +
+                        String.join(",", validBarcodeColumns));
+            }
+
+            validBarcodeColumns.sort((s, t1) -> {
+                int lengthDiff =  s.length() - t1.length();
+                if(lengthDiff == 0) return s.compareTo(t1);
```
Comment (gbggrant) [inline src/main/java/picard/illumina/ExtractBarcodesProgram.java:258]:
```suggestion
                if (lengthDiff == 0) {
                  return s.compareTo(t1);
                }
```

File: src/main/java/picard/illumina/ExtractBarcodesProgram.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+                        messages.add(String.format("Null barcode in column %s of row: %s", header, row.getCurrentLine()));
+                        bcStrings[barcodeNum] = "";
+                    } else {
+                        bcStrings[barcodeNum] = field;
+                    }
+                    ++barcodeNum;
+                }
+                final String bcStr = IlluminaUtil.barcodeSeqsToString(bcStrings);
+                // if the barcode is all Ns don't add it to metrics (we add noCallMetric separately)
+                if(bcStr.contains("N") || bcStr.contains("n")) continue;
```
Comment (gbggrant) [inline src/main/java/picard/illumina/ExtractBarcodesProgram.java:299]:
```suggestion
                if (bcStr.contains("N") || bcStr.contains("n")) {
                  continue;
                }
```
But, we hates 'continue;'

**Review by gbggrant — APPROVED** [review]
A couple of nits.

**Comment by gbggrant** [comment]
@jacarey is this ready to go?

File: src/main/java/picard/illumina/ExtractBarcodesProgram.java (hunk 5 above)
Comment (jacarey) [reply src/main/java/picard/illumina/ExtractBarcodesProgram.java:218]: If it was empty than it wouldn't pass command line validation

```
            if (barcodeToMetrics.keySet().isEmpty()) {
                messages.add("No barcodes have been specified.");
            }
```

**Pull request merged**